enterprise_gateway/services/kernelspecs/handlers.py

```diff
@@ -1,17 +1,175 @@
+def is_kernelspec_model(spec_dict):
+    """Returns True if spec_dict is already in proper form.  This will occur when using a gateway."""
+    return isinstance(spec_dict, dict) and 'name' in spec_dict and 'spec' in spec_dict and 'resources' in spec_dict
+
+
+def kernelspec_model(handler, name, spec_dict, resource_dir):
+    """Load a KernelSpec by name and return the REST API model"""
+    d = {
+        'name': name,
+        'spec': spec_dict,
+        'resources': {}
+    }
+
+    # Add resource files if they exist
+    resource_dir = resource_dir
+    for resource in ['kernel.js', 'kernel.css']:
+        if os.path.exists(pjoin(resource_dir, resource)):
+            d['resources'][resource] = url_path_join(
+                handler.base_url,
+                'kernelspecs',
+                name,
+                resource
+            )
+    for logo_file in glob.glob(pjoin(resource_dir, 'logo-*')):
+        fname = os.path.basename(logo_file)
+        no_ext, _ = os.path.splitext(fname)
+        d['resources'][no_ext] = url_path_join(
+            handler.base_url,
+            'kernelspecs',
+            name,
+            fname
+        )
+    return d
+
 
 # Extends the default handlers from the notebook package with token auth, CORS
 # and JSON errors.
 default_handlers = []
-for path, cls in notebook_handlers.default_handlers:
-    # Everything should have CORS and token auth
+
+
+def exists(obj, chain):
+    _key = chain.pop(0)
+    if _key in obj:
+        return exists(obj[_key], chain) if chain else obj[_key]
+
+
+def apply_user_filter(kernelspec_model, kernel_user=None):
+    if kernel_user:
+        # Check unauthorized list
+        if exists(kernelspec_model, ['spec', 'metadata', 'process_proxy', 'config', 'unauthorized_users']):
+            # Check if kernel_user in kernelspec_model
+            unauthorized_list = kernelspec_model['spec']['metadata']['process_proxy']['config']['unauthorized_users']
+            if kernel_user in unauthorized_list:
+                return None
+        # Now we have check that unauthorized list doesnt exist
+        if exists(kernelspec_model, ['spec', 'metadata', 'process_proxy', 'config', 'authorized_users']):
+            authorized_list = kernelspec_model['spec']['metadata']['process_proxy']['config']['authorized_users']
+            if authorized_list and kernel_user not in authorized_list:
+                return None
+    return kernelspec_model
+
+
+class ModifyKernelSpecHandler(APIHandler):
+    @web.authenticated
+    @gen.coroutine
+    def get(self, kernel_user=None):
+        ksm = self.kernel_spec_manager
+        km = self.kernel_manager
+        model = {}
+        model['default'] = km.default_kernel_name
+        model['kernelspecs'] = specs = {}
+        kspecs = yield maybe_future(ksm.get_all_specs())
+        if kernel_user:
+            self.log.info("Searching kernels for user '%s' " % kernel_user)
+        else:
+            self.log.info("No user. All kernels given")
```

Comment:
Because this is the default, I don't think this should be logged at all.  It can be inferred if the other isn't logged (when DEBUG is enabled).